### DISCUSSION TITLE
Validation resolver cached?

### DISCUSSION THREAD

**Discussion by mortmoe** [react-hook-form/react-hook-form#2259] [Q&A]
I can see in the docs that the resolver should be cached on first render ("This function will be cached inside the custom hook, while context is a mutable object which can be changed on each re-render."). This means if I do this:
```
    const methods = useForm({
        mode: 'onBlur',
        defaultValues: initialValues,
        resolver: yupResolver(validationSchema)
    })
```
... and then "validationSchema" changes from undefined to a yup validator object will it change so validation start working?

The reason I'm asking is that I have a form that I use a jsonSchema to generate, and then I have some functions createYupSchemaFromJsonSchema(schema) and generateDefaultValueFromJsonSchema(schema). The schema is set depending on user interaction meaning that it's not there on the initial render. So... I ran into a problem with the defaultValues not being set and did some digging and found out that I had to use reset like this: 
```
    useEffect(() => {
        methods.reset(initialValues)
    },[initialValues])
```
That fixed the defaultValue problem ... but what confuses me is that the validation works fine. Seems to me like changing the validationSchema (or actually setting it) changes the resolver. I mean... it works, but I don't understand why ... and I don't like things that work without understanding it, they tend to fail eventually :) Can anyone explain why this works?

**Comment by bluebill1049** [comment]
```
useForm({
  context: {xxx},
  resolver: async () => (data, context) => validate.xxx
})
```

i think this is you after?

**Reply by mortmoe** [reply]
Hmm... not sure I get it. I made a little codesandbox to demonstrate:

https://codesandbox.io/s/react-hook-form-demovalidationdefault-ej1md?file=/src/index.js

It will start out with no validationSchema and no defaultValues, then set them both after 10 seconds. This is what happens:
- Before defaultValues and validationSchema is set, the submit-button don't really work
- After 10 seconds validation on submit start working, but not onChange
- If you try to start out with the validationSchema in place... it all seems to work once defaultValues is set (I've noted in the comments what to change)

**Reply by bluebill1049** [reply]
for your usage, you should use context 

`resolver: async (data, context) => console.log(context)`

**Reply by mortmoe** [reply]
@bluebill1049 Thanks for your answer :) I've updated my codesandbox and it pretty much work as expected: 
https://codesandbox.io/s/react-hook-form-demovalidationdefault-ej1md?file=/src/index.js

I ended up creating a dynamicYupResolver like this: 
```
const dynamicYupResolver = async (
  data,
  context,
  validateAllFieldCriteria = false
) => {
  console.log(data, context, validateAllFieldCriteria)
  try {
    let values = data;
    if (context !== undefined) {
      values = await context.validate(data, { abortEarly: false });
    }
    return {
      values,
      errors: {}
    };
  } catch (errors) {
    return {
      values: {},
      errors: transformToNestObject(
        parseErrorSchema(errors, validateAllFieldCriteria)
      )
    };
  }
};
```
However ... I'm not 100% happy with the way context is used here. It's not what it is originally intended for don't you agree? Also, doing it like this we loose the original intention of the context, meaning it can't be used to validate a subset of the schema for example?

**Reply by bluebill1049** [reply]
The context will be injected into resolver in the next release: (if that's what you refer to) https://github.com/react-hook-form/resolvers/pull/30 and the current context is as intended, we are not going to change its behavior.

**Reply by mortmoe** [reply]
Kind of :) I was not implying that the current context is not as intended, but MY use in the function above might not be :)  My issue isn't really with the context ... but that the _schema_ might change _after_ useForm is called.

**Reply by bluebill1049** [reply]
why not keep a set of schemes and use `switch` statements based on the context?

**Reply by mortmoe** [reply]
Hmm... problem is that the schema is dynamically generated based on data from back-end so it might not be available at the time we're calling useForm.

**Reply by mohanaprasadp** [reply]
@bluebill1049 I have a usecase where the validator needs to calculate a constraint dynamically based on another field value. for this, I have used a yup schema where I pass the required value to be added in the constraint like so

```
export const mySchema = maxReceivingWeight => {
	return yup.object().shape({
		weight: yup
			.number('Receiving Weight is not valid')
			.max(
				maxReceivingWeight,
				'Receiving Weight cannot be greater than' + maxReceivingWeight
			)
			.required('Receiving Weight is Required')
	});
};

```
Anyway I can do this?
